notebooks/Model_Training_with_TensorArray.ipynb

```diff
@@ -0,0 +1,5826 @@
+       "351050                False       O     <NA>           O               0   \n",
+       "\n",
+       "                                                embedding  predicted_id  \\\n",
+       "351041 [-2.10291892e-01 -8.53567719e-01  2.75510421e-0...             6   \n",
+       "351042 [-2.32054338e-01 -9.29077148e-01  3.88911217e-0...             6   \n",
+       "351043 [ 0.36844087 -0.68091148 -0.10591034 -0.3448123...             8   \n",
+       "351044 [-3.01310837e-01 -6.54601932e-01 -1.72691718e-0...             8   \n",
+       "351045 [-1.61161020e-01 -6.98910296e-01  2.34246060e-0...             5   \n",
+       "351046 [-5.85671440e-02 -7.95588315e-01  3.36059839e-0...             1   \n",
+       "351047 [ 2.03760266e-01 -7.37309694e-01 -8.88528451e-0...             5   \n",
+       "351048 [-1.03412405e-01 -3.36818755e-01  1.73846602e-0...             1   \n",
+       "351049 [-4.05427039e-01 -6.51651978e-01  2.46960565e-0...             5   \n",
+       "351050 [-1.68292239e-01 -6.47586107e-01  8.14902484e-0...             0   \n",
+       "\n",
+       "       predicted_class predicted_iob predicted_type  \\\n",
+       "351041          I-MISC             I           MISC   \n",
+       "351042          I-MISC             I           MISC   \n",
+       "351043           I-PER             I            PER   \n",
+       "351044           I-PER             I            PER   \n",
+       "351045           I-LOC             I            LOC   \n",
+       "351046           B-LOC             B            LOC   \n",
+       "351047           I-LOC             I            LOC   \n",
+       "351048           B-LOC             B            LOC   \n",
+       "351049           I-LOC             I            LOC   \n",
+       "351050               O             O           None   \n",
+       "\n",
+       "                                       predicted_class_pr  \n",
+       "351041 [1.98323229e-07 1.53130728e-10 2.78405342e-07 2...  \n",
+       "351042 [3.10633651e-05 5.22947228e-06 6.52603219e-11 5...  \n",
+       "351043 [2.00370294e-03 1.00450102e-01 6.38725156e-06 1...  \n",
+       "351044 [1.12122699e-04 3.13711716e-06 1.53931420e-07 4...  \n",
+       "351045 [7.02232647e-03 3.18677296e-03 7.76558124e-08 6...  \n",
+       "351046 [6.36387711e-03 9.02478479e-01 2.26270005e-06 4...  \n",
+       "351047 [2.00492628e-01 5.12452116e-05 1.72780996e-07 8...  \n",
+       "351048 [3.35755616e-02 5.33375311e-01 1.00517707e-07 9...  \n",
+       "351049 [3.03682213e-04 1.33970366e-04 1.83506170e-08 1...  \n",
+       "351050 [9.99999663e-01 3.75752227e-08 1.14784193e-09 1...  "
+      ]
+     },
+     "execution_count": 42,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "# Look at results on the test set\n",
+    "test_results_df = predict_on_df(corpus_df[corpus_df[\"fold\"] == \"test\"], \n",
```

Comment:
These results contain label probabilities for each of the BERT tokens in the test fold of the corpus. To obtain predictions of the named entities in the test fold, you need to convert these probabilities to one "best" label (IOB tag + entity type) per token. The easiest way to get those "best" labels is to take the highest probability label at each token position. Then you 'll want to convert those token labels to (span, entity type pairs), the same way that you processed the corpus itself earlier in this notebook. Once you've made that conversion, you can compute precision and recall a la Analyze_Model_Outputs.ipynb.